File: Replace Words.py
# Time: O(n)
# Space Complexity: O(n)
class Solution:

    # Trie Approach
    # Initialise a Trie
    def __init__(self):
        self.children = {}

    def insert(self, word):
        current = self.children
        for w in word:
            if w not in current:
                current[w] = {}
            current = current[w]
            if '#' in current:
                return
        current['#'] = True

    def search(self, word):
        current = self.children
        for w in word:
            if w not in current:
                return False
            current = current[w]
        return '#' in current

    def startswith(self, word):
        current = self.children
        for w in word:
            if w not in current:
                return False
            current = current[w]
        return True

    def replaceWords(self, dictionary: List[str], sentence: str) -> str:
        # Trie Approach
        result = ''
        sent = sentence.split(" ")
        for word in dictionary:
            self.insert(word)
        # print(self.children)

        for word in sent:
            temp = ''
            curr = self.children
            for ch in word:
                if ch not in curr or '#' in curr:
                    break
                temp += ch
                curr = curr[ch]
            if '#' in curr:
                result += temp
            else:
                result += word
            result += ' '

        return result.strip()


        # # Set Approach (Inefficient)
        # # Time Complexity: O()
        # # Space Complexity: O(n)

        # sent = sentence.split(" ")
        # dic_set = set(dictionary)
        # for i in sent:
        #     flag = False
        #     for j in range(1, len(i)):
        #         # print(i[:j])
        #         if i[:j] in dic_set:

        #             result += i[:j]
        #             flag = True
        #             break
        #     if not flag:
        #         result += i
        #     result += ' '
        # return result.strip()

File: Replace Words.cpp
//Time Complexity- O(n)
//Space Complexity- O(n)

struct trienode{
    
    trienode* children[26];
    string word;
};

trienode* newnode(){
    
    trienode* temp=new trienode();
    for(int i=0;i<26;i++){
        temp->children[i]=NULL;
        temp->word="";
    }
    return temp;
}

void insert(trienode*& root,string &word){
    
    trienode* curr=root;
    for(int i=0;i<word.size();i++){
        int index=word[i]-'a';
        if(!curr->children[index]){
            curr->children[index]=newnode();
        }
        curr=curr->children[index];
    }
    curr->word=word;
}

string searchTrie(trienode* root,string word){
    
    trienode* curr=root;
    for(int i=0;i<word.size();i++){
        int index=word[i]-'a';
        if(!curr->children[index]){
            return "";
        }
        else if(curr->children[index]->word.size()>0){
            return curr->children[index]->word;
        }
        curr=curr->children[index];
    }
    return "";
}

class Solution {
public:
    string replaceWords(vector<string>& dictionary, string sentence) {
        
        trienode* root=newnode();
        string ans="";
        for(string w:dictionary){
            insert(root,w);
        }
        
        stringstream ss(sentence);
        string word;
        while(ss>>word){
            string temp=searchTrie(root,word);
            if(temp.size()>0){
                ans+=temp+' ';
            }
            else{
                ans+=word+' ';
            }
        }
        return ans.substr(0,ans.size()-1);
    }
};